### PR TITLE
Add Kubernetes back-end client

### PR DESCRIPTION
**Pull request by chrisdoherty4**
## Description

Builds on #95 (rebased after PR was created)

This feature builds on #81 and realizes the Hegel portion of the [Kubernetes back-end proposal](https://github.com/tinkerbell/proposals/tree/main/proposals/0026).

Having generalized the client used to acquire metadata, this PR implements a Kubernetes client that leverages an internal client caching mechanism so as to reduce load on the Kubernetes API Server.

Additionally, I put some effort into reworking the Hegel Dockerfile so it could leverage layer caching for Go module dependencies and added an image build command to the Makefile.

The `Watch` API for the client is unimplemented and scheduled for a follow up. This is so we can get off the ground for machine provisioning. The `Watch` API is used for subscription capabilities so doesn't feature in metadata retrieval. 

## How Has This Been Tested?

Built and ran the image with the Kubernetes back-end enabled talking to a KinD cluster containing the necessary CRDs and a `Hardware` resource. 

## How are existing users impacted? What migration steps/scripts do we need?

None. Users enable the Kubernetes back-end by setting the `DATA_MODEL_VERSION=kubernetes` plus some additional env vars.

## Future work

- Implement the `Watch` API for consistency with other back-ends.

### CODE REVIEW

File: hardware/kubernetes.go
Comment (chrisdoherty4) [inline hardware/kubernetes.go:89]: Suggestions on what to do here as its not true. Initially I thought a mutex protected `k.close` then set `k.close = nil` in the `Close()` func with a `return k.close != nil` for `IsHealthy()`. @micahhausler is there a better way in the manager to check if we're 'connected'?

File: hardware/client.go
Comment (micahhausler) [inline hardware/client.go:91]: I'm trying to standardize components on the same flags. Can we go with flags for `kubernetes-api` and `kubeconfig` and then ingest with viper here?

File: hardware/kubernetes.go
Comment (micahhausler) [reply hardware/kubernetes.go:89]: It doesn't look like controller-runtime's client exposes any health metrics.I think this is good enough

File: hardware/kubernetes.go
Comment (micahhausler) [inline hardware/kubernetes.go:119]: Why is watch not implemented here?

File: hardware/kubernetes.go
Comment (micahhausler) [inline hardware/kubernetes.go:127]: What structure does Export expect here? Is the CRD spec going to be ingested correctly?

File: hardware/kubernetes_test.go
Comment (micahhausler) [inline hardware/kubernetes_test.go:15]: Can we use envtest instead of a hardcoded kubeconfig?

**Review by micahhausler — CHANGES_REQUESTED** [review]
A few questions, and some cleanup.

File: hardware/kubernetes_test.go
Comment (chrisdoherty4) [reply hardware/kubernetes_test.go:15]: This is a local test I put together for ease. Its going to be replaced entirely with unit tests and e2e tests. 

The e2e tests should address the [export comment](https://github.com/tinkerbell/hegel/pull/82#discussion_r851402168).

File: hardware/kubernetes.go
Comment (chrisdoherty4) [reply hardware/kubernetes.go:127]: [See my other reply](https://github.com/tinkerbell/hegel/pull/82#discussion_r851406714). 

I think there are some inconsistencies. There will likely be a need to translate between the CRD spec and the export expectations but some fields match up already hence I was able to validate things "work".

File: hardware/kubernetes.go
Comment (chrisdoherty4) [reply hardware/kubernetes.go:119]: `Watch` is only used by the gRPC service. I've been told its used by Equinix Metal but the metadata capability doesn't need it, I don't think, hence I've left it out of this PR to be implemented in a follow up.

The changes in this PR should get us off the ground with respect to machine provisioning.

File: hardware/client.go
Comment (chrisdoherty4) [reply hardware/client.go:91]: Hegel doesn't support CLI flags at present hence I purposely avoided adding anything for this PR. 

I'd like to address it separately because I think it'd be worth introducing CLI arguments holistically; does that work for you?

File: .gitattributes
Comment (chrisdoherty4) [inline .gitattributes:1]: This won't take effect until merged but it'll exclude `go.sum` from the loc on PRs.

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [inline hardware/kubernetes_client.go:175]: Unit test

File: hardware/kubernetes.go
Comment (chrisdoherty4) [reply hardware/kubernetes.go]: I've encoded the minimal expectations into a new export construct to be safe (there were some problems anyhow so this fixes them). 

It would be useful to move away entirely from this approach but it likely means breaking compatibility with the custom metadata rules that can be added via an env var on startup. I suspect others are using that capability which makes it difficult to detach from the jq parsing approach and head toward something that's more clearly defined.

File: .golangci.yml
Comment (displague) [inline .golangci.yml:36]: This sounds like the default. Any reason to set it explicitly?
https://github.com/polyfloyd/go-errorlint/blob/master/README.md#examples

File: go.mod
Comment (displague) [inline go.mod:36]: What PR's do we depend on to remove this `replace`?

File: hardware/kubernetes_client.go
Comment (displague) [inline hardware/kubernetes_client.go:230]: Seems we could be using the existing types for most of this to prevent introducing diff and drift:

https://github.com/tinkerbell/tink/blob/main/pkg/apis/core/v1alpha1/hardware_types.go#L156-L162

File: hardware/kubernetes_client_test.go
Comment (displague) [inline hardware/kubernetes_client_test.go:64]: Would a short string comparison should work for this test?

File: hardware/kubernetes_client.go
Comment (displague) [inline hardware/kubernetes_client.go:172]: This name (`ExportableK8sHardwareHardwareFromK8sObject`) suggests to me that it could be defined, instead, as:

```go
type K8sHardware struct {}
func (*K8sHardware) FromTinkHardware(*tinkv1alpha1.Hardware)
func (*K8sHardware) Export() ([]byte, error)
```

I kind of wish we could avoid the `Export` name and go with `MarshalJSON` to fulfill `Marshaler`. I didn't look around to see how many places would be affected and whether this would be a breaking change, but I do see that at least one `Export` invocation preexists.

**Review by displague — COMMENTED** [review]
I took an initial look to get a sense of what's in flux and left a few questions.

File: .gitattributes
Comment (chrisdoherty4) [reply .gitattributes:1]: Do you mean `go.mod`? `go.sum` is generated and just a bunch of checksums based on the `go.mod` dependency tree.

Comment (chrisdoherty4) [reply .gitattributes:1]: If we don't want this I can remove it. I've done this on past projects to remove misleading line counts.

File: .golangci.yml
Comment (chrisdoherty4) [reply .golangci.yml:36]: The root project does, but `golangci-lint` has it on by default. See https://golangci-lint.run/usage/linters/#errorlint. I had Github action errors as a result.

File: go.mod
Comment (chrisdoherty4) [reply go.mod:36]: https://github.com/tinkerbell/tink/pull/573

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client.go:230]: The existing types have a different structure so while they seem similar, they don't satisfy the Hegel metadata expectations from the `Export()` call. 

Moreover, I think its beneficial to have a clear distinction between Hegel metadata structure and back-end storage format as it allows us to change the structure of the back-end data without impacting hegel given there's an explicit translation.

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client.go:172]: Can use `K8sHardware`. I'm not sure that method should be on the struct though. It happens to use the fields but its not a behavior the struct should provide.

Comment (chrisdoherty4) [reply hardware/kubernetes_client.go]: Renamed. Re-open if you'd like to talk methods more.

File: go.mod
Comment (chrisdoherty4) [reply go.mod:36]: I can't add labels but if you could add a do-not-merge I'll track the other and ping when its merged.

File: .golangci.yml
Comment (mmlb) [reply .golangci.yml:36]: This file is managed centrally via github.com/tinkerbell/lint-install in an effort to have all the go code passing the same linter settings and thus hopefully look/feel similar. If we're to disable this rule we should do that there. I'm not sure I agree about disabling this btw, if you do send this change over there please explain further how this ends up displaying Hyrums Law.

File: go.mod
Comment (mmlb) [reply go.mod:36]: I've added the label.

File: .golangci.yml
Comment (chrisdoherty4) [reply .golangci.yml:36]: Errors are part of the package API. That means if you expose a Kubernetes specific error the Kubernetes specific error is part of the packages API whether you want it or not. That should really be a conscious decision by the package author, not mandated by a linter.

Hyrum's law comes into play because consumers become dependent on that error type and that was never intended making things harder to change in the future.

It's less relevant for internal packages but I typically advocate for the rationale above anyway.

Comment (chrisdoherty4) [reply .golangci.yml:36]: Fwiw errorlint even calls this out as a trade off.

File: hardware/kubernetes_client_test.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client_test.go:64]: I suspect so. I'll put something together.

File: .gitattributes
Comment (displague) [reply .gitattributes:1]: I see. `k8s.io/client-go` and `sigs.k8s.io/controller-runtime` were added while a few other dependencies were removed. Quite a large number of indirect dependencies came from this, but I think these shouldn't be problems because Tink has roughly the same dependencies (differing on version).

Comment (chrisdoherty4) [reply .gitattributes:1]: Yes. I was alarmed myself at the transitive dependencies introduced because it literally doubled but then realized it was mostly due to Kube pkgs.

File: hardware/kubernetes_client_test.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client_test.go:65]: Added

File: hardware/kubernetes.go
Comment (chrisdoherty4) [reply hardware/kubernetes.go]: Resolving as I don't intend to address. Reopen if needed.

File: hardware/kubernetes_test.go
Comment (chrisdoherty4) [reply hardware/kubernetes_test.go]: Won't address on this PR. Reopen if needed.

File: .golangci.yml
Comment (chrisdoherty4) [reply .golangci.yml]: Removed and raised in the `lint-install` repo.

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client.go]: I think this is OK. Please unresolve if needed.

File: Makefile
Comment (micahhausler) [inline Makefile:3]: Needs a `.PHONY`

File: Makefile
Comment (micahhausler) [inline Makefile:20]: Needs a `.PHONY`

File: hardware/kubernetes_client.go
Comment (micahhausler) [inline hardware/kubernetes_client.go:203]: Nit: comment doesn't match struct name

**Review by micahhausler — APPROVED** [review]
LGTM modulo minor nits, and pending merge of tinkerbell/tink#573

File: go.mod
Comment (micahhausler) [inline go.mod:40]: @chrisdoherty4 tinkerbell/tink#573 is merged!

File: go.mod
Comment (micahhausler) [inline go.mod:29]: Why is the k8s stuff in a separate require block? Is that autogenerated?

Comment (chrisdoherty4) [reply go.mod:29]: I honestly don't know. I've never seen `go` do this before. The indirect deps appearing separate was an addition in 1.17 I believe but this is peculiar.

Comment (chrisdoherty4) [reply go.mod:29]: It might have been merge conflict stuff actually.

File: hardware/kubernetes_client.go
Comment (jacobweinstock) [inline hardware/kubernetes_client.go:55]: If this err doesnt manifest until some set of circumstances that are only satisfied when an incoming request is handled, this might be difficult to troubleshoot. Having this called in func main somehow, might be something to consider.

File: hardware/kubernetes_client.go
Comment (jacobweinstock) [inline hardware/kubernetes_client.go:53]: hmmm...removing this goroutine from this function might be something to consider. The name suggests that it is a constructor and having a constructor start a goroutine seems a bit unusual, could be just me though.

refs: [1](https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions), [2](https://github.com/golang/go/wiki/CodeReviewComments#goroutine-lifetimes)

File: hardware/kubernetes_client.go
Comment (jacobweinstock) [inline hardware/kubernetes_client.go:167]: with the input to this func being `*tinkv1alpha1.Hardware` does it need the `FromK8sObject` in its name?

[Go and short names](https://github.com/golang/go/wiki/CodeReviewComments#variable-names)

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client.go:55]: Now that I've plumbed in `oklog/run` I can do this. When it was originally possible this was nasty to do :(

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client.go]: Adjusted.

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client.go:53]: Generally I agree. The way the app is designed though makes this difficult to do. The `NewClient()` call is a factory function returning an interface so anything that needs to be done for a given implementation _must_ be done during that call. Without re-designing this factory I don't think there's much I can do.

File: hardware/kubernetes_client.go
Comment (chrisdoherty4) [reply hardware/kubernetes_client.go:55]: Scratch that. See my [other comment](https://github.com/tinkerbell/hegel/pull/82#discussion_r867362342).

Resolving this in favor of that discussion.

**Pull request merged**
